File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.java

import java.util.*;

public class Main {
        public static void main(String[] args) {
            Scanner scanner = new Scanner(System.in);

            int N = scanner.nextInt();
            int M = scanner.nextInt();
            int R = scanner.nextInt();

            int[][] matrix = new int[N][M];
            int[][] answer = new int[N][M];

            for (int r = 0; r < N; r++) {
                for (int c = 0; c < M; c++) {
                    matrix[r][c] = scanner.nextInt();
                }
            }

            ArrayList<Integer> dq = new ArrayList<>();

            int l = Math.min(M, N);

            for (int i = 0; i < l / 2; i++) {
                for (int c = i; c < M - i; c++) {
                    dq.add(matrix[i][c]);
                }

                for (int r = i; r < N - i; r++) {
                    if (r != i) {
                        dq.add(matrix[r][M - i - 1]);
                    }
                }

                for (int c = M - i - 1; c > i; c--) {
                    if (c != M - i - 1) {
                        dq.add(matrix[N - i - 1][c]);
                    }
                }

                for (int r = N - i - 1; r > i; r--) {
                    if (r != N - i - 1 || r != i - 1) {
                        dq.add(matrix[r][i]);
                    }
                }

                Collections.rotate(dq, -R);


                // 껍질 깎았던 방법 그대로 정답에 채워넣기

                    for (int c = i; c < M - i; c++) {
                        answer[i][c] = dq.get(0);
                        dq.remove(0);
                    }

                    for (int r = i; r < N - i; r++) {
                        if (r != i) {
                            answer[r][M - i - 1] = dq.get(0);
                            dq.remove(0);
                        }
                    }

                    for (int c = M - i - 1; c > i; c--) {
                        if (c != M - i - 1) {
                            answer[N - i - 1][c] = dq.get(0);
                            dq.remove(0);
                        }
                    }

                    for (int r = N - i - 1; r > i; r--) {
                        if (r != N - i - 1 || r != i - 1) {
                            answer[r][i] = dq.get(0);
                            dq.remove(0);
                        }
                    }

            }

            for (int[] row : answer) {
                for (int val : row) {
                    System.out.print(val + " ");
                }
                System.out.println();
            }

    }
}

File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.py
from collections import deque

# 첫 행,열 & 마지막 행,열 

N,M,R = map(int,input().split())
matrix = []
answer = [[0 for _ in range(M)] for _ in range(N)]
for r in range(N):
    matrix.append(list(map(int,input().split())))

rstart = 0
rend = N
cstart = 0
cend = M

dq = deque()

l = min(M,N) # 주의 # 껍데기 갯수는 N과 M중 작은 것을 기준으로 정해짐

for i in range(l//2):
    for c in range(i, M-i):
        dq.append(matrix[i][c]) # 맨위 ->

    for r in range(i , N-i):
        if(r != i): # 맨 오른쪽 모서리 중복처리
            dq.append(matrix[r][M-i-1]) # 맨 오른쪽

    for c in range(M- i-1, i,-1):
        if(c != M-i-1):
            dq.append(matrix[N- i-1][c]) # 맨 아래  <-

    for r in range(N- i-1, i ,-1):
        if(r != N-i-1 or r != i-1):
            dq.append(matrix[r][i]) #맨 왼쪽 

    dq.rotate(-R)       

    for c in range(cstart+i, cend-i):
        answer[rstart + i][c] = dq.popleft() # 맨위 ->

    for r in range(rstart+i , rend-i):
        if(r != rstart+i): # 맨 오른쪽 모서리 중복처리
            answer[r][cend - i-1] = dq.popleft() # 맨 오른쪽

    for c in range(cend - i-1, cstart+i,-1):
        if(c != cend -i-1):
            answer[rend - i-1][c] = dq.popleft() # 맨 아래  <-

    for r in range(rend - i-1, rstart+i ,-1):
        if(r != rend -i-1 or r != rstart-1):
            answer[r][cstart +i] = dq.popleft() #맨 왼쪽 

for r in answer:
    print(*r)